fairscale/experimental/wgit/sha1_store.py

```diff
@@ -4,160 +4,257 @@
                 sys.exit(ExitCode.FILE_EXISTS_ERROR)
+            # Create a new json file for this new store.
+            self._store_json_dict()
 
-    def add(self, file_path: Path, parent_sha1: str) -> str:
+        # This is an internal error since caller of this our own wgit code.
+        assert self._path.exists(), "SHA1 store does not exist and init==False"
+
+        # Init temp dir.
+        if tmp_dir:
+            # Caller supplied tmp dir
+            assert Path(tmp_dir).is_dir(), "incorrect input"
+            self._tmp_dir = Path(tmp_dir)
+        else:
+            # Default tmp dir, need to clean it.
+            self._tmp_dir = self._path.joinpath("tmp")
+            shutil.rmtree(self._tmp_dir, ignore_errors=True)
+            self._tmp_dir.mkdir()
+
+    def _load_json_dict(self) -> None:
+        """Loading json dict from disk."""
+        with open(self._ref_file_path, "r") as f:
+            self._json_dict = json.load(f)
+
+    def _store_json_dict(self) -> None:
+        """Storing json dict to disk."""
+        with open(self._ref_file_path, "w", encoding="utf-8") as f:
+            json.dump(self._json_dict, f, ensure_ascii=False, indent=4)
+
+    def add(self, file_or_obj: Union[Path, Tensor, OrderedDict]) -> str:
         """
-        Adds a file/checkpoint to the internal sha1_store and the sha1 references accordingly.
-        First, a sha1 hash is calculated. Utilizing the sha1 hash string, the actual file in <in_file_path> is moved
-        within the sha1_store and the sha1 reference file is updated accordingly with the information of their parents
-        node (if exists) and whether the new version is a leaf node or not.
+        Adds a file/object to the internal sha1_store and the sha1 references
+        accordingly.
+
+        First, a sha1 hash is calculated. Utilizing the sha1 hash string, the actual file
+        in <file_or_obj> is moved within the sha1_store and the reference file is updated.
+        If the input is an object, it will be store in the self._tmp_dir and then moved.
 
         Args:
-            in_file_path (str): path to the file to be added to the sha1_store.
+            file_or_obj (str or tensor or OrderedDict):
+                Path to the file to be added to the sha1_store or an in-memory object
+                that can be handled by torch.save.
         """
+        # Use `isinstance` not type() == Path since pathlib returns OS specific
+        # Path types, which inherit from the Path class.
+        if isinstance(file_or_obj, (Path, str)):
+            # Make sure it is a valid file.
+            torch.load(cast(Union[Path, str], file_or_obj))
+            file_path = Path(file_or_obj)
+            remove_tmp = False
+        elif isinstance(file_or_obj, (Tensor, OrderedDict)):
+            # Serialize the object into a tmp file.
+            file_path = self._get_tmp_file_path()
+            torch.save(cast(Union[Tensor, OrderedDict], file_or_obj), file_path)
```

Comment:
Is this casting for adhering to mypy type checking?

Reply:
Exactly😀